extensions/hdfs-storage/src/main/java/io/druid/storage/hdfs/HdfsDataSegmentFinder.java

```diff
@@ -0,0 +1,110 @@
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package io.druid.storage.hdfs;
+
+import com.fasterxml.jackson.databind.ObjectMapper;
+import com.google.common.collect.Sets;
+import com.google.inject.Inject;
+import com.metamx.common.logger.Logger;
+import io.druid.segment.loading.DataSegmentFinder;
+import io.druid.segment.loading.SegmentLoadingException;
+import io.druid.timeline.DataSegment;
+import org.apache.hadoop.conf.Configuration;
+import org.apache.hadoop.fs.FileSystem;
+import org.apache.hadoop.fs.LocatedFileStatus;
+import org.apache.hadoop.fs.Path;
+import org.apache.hadoop.fs.RemoteIterator;
+
+import java.io.IOException;
+import java.util.Map;
+import java.util.Set;
+
+/**
+ */
+public class HdfsDataSegmentFinder implements DataSegmentFinder
+{
+
+  private static final Logger log = new Logger(HdfsDataSegmentFinder.class);
+
+  private final Configuration config;
+  private final ObjectMapper mapper;
+
+  @Inject
+  public HdfsDataSegmentFinder(Configuration config, ObjectMapper mapper)
+  {
+    this.config = config;
+    this.mapper = mapper;
+  }
+
+  @Override
+  public Set<DataSegment> findSegments(String workingDirPathStr, boolean updateDescriptor)
+      throws SegmentLoadingException
+  {
+    final Set<DataSegment> segments = Sets.newHashSet();
+    final Path workingDirPath = new Path(workingDirPathStr);
+    FileSystem fs;
+    try {
+      fs = workingDirPath.getFileSystem(config);
+
+      if (!fs.exists(workingDirPath)) {
+        throw new SegmentLoadingException("Working directory [%s] doesn't exist.", workingDirPath);
+      }
+
+      if (!fs.isDirectory(workingDirPath)) {
+        throw new SegmentLoadingException("Working directory [%s] is not a directory!?", workingDirPath);
+      }
+
+      final RemoteIterator<LocatedFileStatus> it = fs.listFiles(workingDirPath, true);
+      while (it.hasNext()) {
+        final LocatedFileStatus locatedFileStatus = it.next();
+        final Path path = locatedFileStatus.getPath();
+        if (path.getName().equals("descriptor.json")) {
+          final Path indexZip = new Path(path.getParent(), "index.zip");
+          if (fs.exists(indexZip)) {
+            final DataSegment dataSegment = mapper.readValue(fs.open(path), DataSegment.class);
+            final Map<String, Object> loadSpec = dataSegment.getLoadSpec();
+            if (!loadSpec.get("type").equals(HdfsStorageDruidModule.SCHEME) || !loadSpec.get("path")
+                                                                                        .equals(indexZip.toString())) {
+              loadSpec.put("type", HdfsStorageDruidModule.SCHEME);
+              loadSpec.put("path", indexZip.toString());
```

Comment:
indexZip.toString() would bring `hdfs://host:port` as well?  I would just put the absolute path.